core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -2422,18 +2422,16 @@ public RelBuilder join(JoinRelType joinType, RexNode condition,
     }
     if (correlate) {
       final CorrelationId id = Iterables.getOnlyElement(variablesSet);
-      if (!RelOptUtil.notContainsCorrelation(left.rel, id, Litmus.IGNORE)) {
-        throw new IllegalArgumentException("variable " + id
-            + " must not be used by left input to correlation");
-      }
       // Correlate does not have an ON clause.
       switch (joinType) {
       case LEFT:
       case SEMI:
       case ANTI:
         // For a LEFT/SEMI/ANTI, predicate must be evaluated first.
         stack.push(right);
-        filter(condition.accept(new Shifter(left.rel, id, right.rel)));
+        filter(
+            RelOptUtil.correlateLeftShiftRight(getRexBuilder(),
+                left.rel, id, right.rel, condition));
```

Comment:
For quite some time (since 2016), this API is used to create correlated joins by passing correlated variables in the parameters of the method without requiring that the right side already has the correlated condition. I am not sure that removing the correlation altogether (when there is no correlated condition in the right) is a desirable behavior. I am a bit afraid that people rely on the existing behavior and changing it might break other projects.  

To put it a bit differently if the plans emitted by this method are not semantically equivalent before and after these changes we may be in trouble. Have you considered examples where the left/right input has some sort/limit in it? Would replacing the correlate with join still lead to the same results?

Reply:
@zabetak, I believe you are conflating limitations of the current decorrelator and being semantically equivalent.  Sort and limit nodes can be in correlated nodes and the decorrelator could decorrelate them as long as their are no correlated variables in the limit node and it children, and the sort node only requires the correlated variables not occur in its child nodes because any correlated variable that occurs in a sort node could simple be written out.

Various other trivial simplification are performed through out the RelBuilder, and this does not appear to be out of character of those simplifications.  I am hard pressed to imagine any circumstance where this would cause a breaking change out side of the expectation the this call returns a correlated node, which in that case they are not programing against the Apache Calcite public api, and the onerous is on them, not calcite.

Reply:
I guess we both agree that `Correlate` and `Join` are not equivalent in general. With these changes though a plan that previously used a `Correlate` may now use a `Join` thus I would like to make sure that this is valid in every case (no matter how the left or right relation in the stack looks like).

Reply:
@zabetak, I agree a `Correlate` and `Join` are not generally equivalent.  However, in the trivial case where their are not Correlated Variables, they are equivalent.  

The documentation for `Correlate` states:
`It behaves like a kind of {@link org.apache.calcite.rel.core.Join},
 but works by setting variables in its environment and restarting its
 right-hand input.`
 
So when the second part is a No Op, then it behaves like a join with no exceptions.

I am not following your second statement, are you concerned that an exception is no long thrown when a correlated id is passed into an outer join where the join is not actually correlated?  Or are you looking for an exhaustive that every node is equivalent wether a join or correlate is on top?  This covered by nodes being independent/isolated from their parents, which they are not if they contain a correlated variable, but that has already been established by the call `return !RelOptUtil.correlationColumns(..., rightRel).isEmpty();`.   Or are you concerned that `RelOptUtil.correlationColumns` is bugged?

Reply:
@zabetak, I removed the check right side for considering if the join is correlated, would this pr be acceptatable?

Reply:
Let's consider for instance the following plans which can be produced with the code before and after the changes in this PR.

```
LogicalProject(DEPTNO=[$0], EMPNO=[$3])
  LogicalCorrelate(correlation=[$cor0], joinType=[left], requiredColumns=[{0}])
    LogicalTableScan(table=[[scott, DEPT]])
    LogicalFilter(condition=[=($cor0.DEPTNO, $7)])
      LogicalSort(sort0=[$5], dir0=[ASC], fetch=[2])
        LogicalTableScan(table=[[scott, EMP]])

LogicalProject(DEPTNO=[$0], EMPNO=[$3])
  LogicalJoin(condition=[=($0, $10)], joinType=[left])
    LogicalTableScan(table=[[scott, DEPT]])
    LogicalSort(sort0=[$5], dir0=[ASC], fetch=[2])
      LogicalTableScan(table=[[scott, EMP]])
```
What I would like to make sure is that whatever is under the right side of the join/correlate+filter  does not give different results if we switch from correlate+filter to join. I need to find some time to think on it a bit more.

Reply:
I think the change is indeed valid and useful. In terms of process though I think this commit (change in {{RelBuilder}} + tests) deserves its own JIRA (and to facilitate things its own PR). This will give more visibility to the change (release notes etc), and can be merged (or reverted if needed in the future) independently of the other changes. 

Can you please create a JIRA about it with a clear summary and description of the change? Also for the existing tests I don't think you need/should modify them so that they generate a `Correlate`; just update the plans and add extra tests if you deem necessary.